File: src/main/java/org/jukeboxmc/blockentity/BlockEntityBlastFurnace.java
package org.jukeboxmc.blockentity;

import com.nukkitx.nbt.NbtMap;
import com.nukkitx.nbt.NbtMapBuilder;
import com.nukkitx.nbt.NbtType;
import org.jetbrains.annotations.NotNull;
import org.jukeboxmc.block.Block;
import org.jukeboxmc.block.direction.BlockFace;
import org.jukeboxmc.inventory.BlastFurnaceInventory;
import org.jukeboxmc.inventory.InventoryHolder;
import org.jukeboxmc.item.Item;
import org.jukeboxmc.math.Vector;
import org.jukeboxmc.player.Player;

import java.util.ArrayList;
import java.util.List;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class BlockEntityBlastFurnace extends SmeltingComponent implements InventoryHolder {

    private final @NotNull BlastFurnaceInventory blastFurnaceInventory;

    public BlockEntityBlastFurnace(@NotNull Block block, BlockEntityType blockEntityType ) {
        super( block, blockEntityType );
        this.blastFurnaceInventory = new BlastFurnaceInventory( this );
        this.initInventory( this.blastFurnaceInventory );
    }

    @Override
    public boolean interact(@NotNull Player player, @NotNull Vector blockPosition, Vector clickedPosition, BlockFace blockFace, Item itemInHand ) {
        player.openInventory( this.blastFurnaceInventory, blockPosition );
        return super.interact( player, blockPosition, clickedPosition, blockFace, itemInHand );
    }

    @Override
    public void fromCompound(@NotNull NbtMap compound ) {
        super.fromCompound( compound );

        List<NbtMap> items = compound.getList( "Items", NbtType.COMPOUND );
        for ( NbtMap nbtMap : items ) {
            Item item = this.toItem( nbtMap );
            byte slot = nbtMap.getByte( "Slot", (byte) 127 );
            if ( slot == 127 ) {
                this.blastFurnaceInventory.addItem( item, false );
            } else {
                this.blastFurnaceInventory.setItem( slot, item, false );
            }
        }
    }

    @Override
    public @NotNull NbtMapBuilder toCompound() {
        NbtMapBuilder builder = super.toCompound();

        List<NbtMap> itemsCompoundList = new ArrayList<>();
        for ( int slot = 0; slot < this.blastFurnaceInventory.getSize(); slot++ ) {
            NbtMapBuilder itemCompound = NbtMap.builder();
            Item item = this.blastFurnaceInventory.getItem( slot );

            itemCompound.putByte( "Slot", (byte) slot );
            this.fromItem( item, itemCompound );

            itemsCompoundList.add( itemCompound.build() );
        }

        builder.putList( "Items", NbtType.COMPOUND, itemsCompoundList );
        return builder;
    }

    public @NotNull BlastFurnaceInventory getBlastFurnaceInventory() {
        return this.blastFurnaceInventory;
    }
}

File: src/main/java/org/jukeboxmc/blockentity/BlockEntityBlastFurnace.kt
package org.jukeboxmc.blockentity

import org.cloudburstmc.nbt.NbtMap
import org.cloudburstmc.nbt.NbtMapBuilder
import org.cloudburstmc.nbt.NbtType
import org.jukeboxmc.block.Block
import org.jukeboxmc.block.direction.BlockFace
import org.jukeboxmc.inventory.BlastFurnaceInventory
import org.jukeboxmc.inventory.InventoryHolder
import org.jukeboxmc.item.Item
import org.jukeboxmc.math.Vector
import org.jukeboxmc.player.Player

/**
 * @author LucGamesYT
 * @version 1.0
 */
class BlockEntityBlastFurnace(
    block: Block,
    blockEntityType: BlockEntityType,
) : SmeltingComponent(block, blockEntityType), InventoryHolder {
    val blastFurnaceInventory: BlastFurnaceInventory = BlastFurnaceInventory(this)

    init {
        initInventory(blastFurnaceInventory)
    }

    override fun interact(
        player: Player,
        blockPosition: Vector,
        clickedPosition: Vector?,
        blockFace: BlockFace?,
        itemInHand: Item?,
    ): Boolean {
        player.openInventory(blastFurnaceInventory, blockPosition)
        return super.interact(player, blockPosition, clickedPosition, blockFace, itemInHand)
    }

    override fun fromCompound(compound: NbtMap) {
        super.fromCompound(compound)
        val items = compound.getList("Items", NbtType.COMPOUND)
        for (nbtMap in items) {
            val item = toItem(nbtMap)
            val slot = nbtMap.getByte("Slot", 127.toByte())
            if (slot.toInt() == 127) {
                blastFurnaceInventory.addItem(item, false)
            } else {
                blastFurnaceInventory.setItem(slot.toInt(), item, false)
            }
        }
    }

    override fun toCompound(): NbtMapBuilder {
        val builder = super.toCompound()
        val itemsCompoundList: MutableList<NbtMap> = mutableListOf()
        for (slot in 0 until blastFurnaceInventory.size) {
            val itemCompound = NbtMap.builder()
            val item = blastFurnaceInventory.getItem(slot)
            itemCompound.putByte("Slot", slot.toByte())
            fromItem(item, itemCompound)
            itemsCompoundList.add(itemCompound.build())
        }
        builder.putList("Items", NbtType.COMPOUND, itemsCompoundList)
        return builder
    }
}
